### PR TITLE
[Performance][All] Avoiding copies in ranges

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

Avoiding copies in ranges for performance reasons

**🆕 Changelog**

- [Avoiding copies in ranges in core](https://github.com/KratosMultiphysics/Kratos/commit/c701e947350e6112a424513ac50be2b5d69d4451)
- [Avoiding copies in ranges in ContactApp](https://github.com/KratosMultiphysics/Kratos/commit/89c8790e4dbda129e815981e750d6ff82afcb107)
- [Avoiding copies in ranges in FluidApp](https://github.com/KratosMultiphysics/Kratos/commit/feb9d6b8b7ab41be5932743c6ace3096b3299f8f)
- [Avoiding copies in ranges in MeshingApp](https://github.com/KratosMultiphysics/Kratos/commit/ad2f69b527df71774987ee868bf0df9c61cd2493)
- [Avoiding copies in ranges in StructuralApp](https://github.com/KratosMultiphysics/Kratos/commit/6a1299635ec139b2fea1ca2215e11b9a7282ce47)

### CODE REVIEW

File: kratos/processes/tetrahedral_mesh_orientation_check.cpp [hunk 1]
```diff
@@ -136,15 +136,15 @@ void TetrahedralMeshOrientationCheck::Execute()
...
+                    for (const Condition::Pointer& p_cond : list_conditions) {
                         p_cond->Set(VISITED,true);
                     }
 
                     if(mrOptions.Is(ASSIGN_NEIGHBOUR_ELEMENTS_TO_CONDITIONS)) {
                         GlobalPointersVector< Element > vector_of_neighbours;
                         vector_of_neighbours.resize(1);
                         vector_of_neighbours(0) = Element::WeakPointer( *it_elem.base() );
-                        for (Condition::Pointer p_cond : list_conditions) {
+                        for (const Condition::Pointer& p_cond : list_conditions) {
```
Comment (AlejandroCornejo) [inline kratos/processes/tetrahedral_mesh_orientation_check.cpp:147]: so in here we were doing sa copy of the pointers?

**Review by AlejandroCornejo — DISMISSED** [review]

File: kratos/processes/tetrahedral_mesh_orientation_check.cpp (hunk 1 above)
Comment (loumalouomega) [reply kratos/processes/tetrahedral_mesh_orientation_check.cpp:147]: Of smart pointers in this case. But now I am thinking that the Condition::Pointer is an intrusive_ptr... Maybe we can keep it as it is

Comment (loumalouomega) [reply kratos/processes/tetrahedral_mesh_orientation_check.cpp:147]: I will wait to @RiccardoRossi to confirm if this improves (I would say yes)

Comment (RiccardoRossi) [reply kratos/processes/tetrahedral_mesh_orientation_check.cpp:147]: copying inteueive_ptrs is as expensive as for nor al smart_ptrs, so i wpuld expect this imprpves...nevertheless please give me the time to have a look

Comment (loumalouomega) [reply kratos/processes/tetrahedral_mesh_orientation_check.cpp:147]: @RiccardoRossi can I proceed with the merge?

**Review by roigcarlo — COMMENTED** [review]
Just commenting because I am not familiar with most of the code. To me looks good.
That being said the changes for loops that act on iterators, ptr and basic types probably wont affect the performance, but I understand we want to keep them consistent through the code so :+1:

File: kratos/processes/tetrahedral_mesh_orientation_check.cpp (hunk 1 above)
Comment (loumalouomega) [reply kratos/processes/tetrahedral_mesh_orientation_check.cpp:147]: @RiccardoRossi can I proceed?

Comment (loumalouomega) [reply kratos/processes/tetrahedral_mesh_orientation_check.cpp:147]: @RiccardoRossi ping

File: applications/ContactStructuralMechanicsApplication/custom_processes/base_contact_search_process.cpp [hunk 2]
```diff
@@ -744,7 +744,7 @@ void BaseContactSearchProcess<TDim, TNumNodes, TNumNodesMaster>::SearchUsingOcTr
                 // If not active we check if can be potentially in contact
                 if (mCheckGap == CheckGap::MappingCheck) {
                     for (auto p_leaf : leaves) {
-                        for (auto p_cond_master : *(p_leaf->pGetObjects())) {
+                        for (const auto& p_cond_master : *(p_leaf->pGetObjects())) {
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_processes/base_contact_search_process.cpp:747]: ok, however i would say that you were only making a copy of a pointer

File: applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_weighted_gap.cpp [hunk 3]
```diff
@@ -296,8 +296,8 @@ namespace Kratos
 
             // We set the database
             ModelPart& r_computing_contact_model_part = rModelPart.GetSubModelPart("ComputingContact");
-            for (auto p_slave_cond : slave_conds) {
-                for (auto p_master_cond : master_conds) {
+            for (const auto& p_slave_cond : slave_conds) {
+                for (const auto& p_master_cond : master_conds) {
                     id_cond++;
                     const PairedCondition& r_reference_condition = dynamic_cast<const PairedCondition&>(KratosComponents<Condition>::Get("ALMFrictionalMortarContactCondition3D4N"));
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_weighted_gap.cpp:302]: beware, this Get is not fast. You should rather get it once and then clone it as you need

Comment (RiccardoRossi) [reply applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_weighted_gap.cpp:302]: (also i don't really understand what you are doing here)

File: applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_weighted_gap.cpp [hunk 4]
```diff
@@ -711,8 +711,8 @@ namespace Kratos
             // We set the database
             const auto& r_process_info = rModelPart.GetProcessInfo();
             ModelPart& r_computing_contact_model_part = rModelPart.GetSubModelPart("ComputingContact");
-            for (auto p_slave_cond : slave_conds) {
-                for (auto p_master_cond : master_conds) {
+            for (const auto& p_slave_cond : slave_conds) {
+                for (const auto& p_master_cond : master_conds) {
                     id_cond++;
                     const PairedCondition& r_reference_condition = dynamic_cast<const PairedCondition&>(KratosComponents<Condition>::Get("ALMFrictionalMortarContactCondition3D4N"));
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_weighted_gap.cpp:717]: if you seek performance do the get once and reuse it

File: applications/FluidDynamicsApplication/custom_processes/apply_compressible_navier_stokes_boundary_conditions_process.cpp [hunk 5]
```diff
@@ -29,12 +29,12 @@ ApplyCompressibleNavierStokesBoundaryConditionsProcess::ApplyCompressibleNavierS
     mSubsonicBCs.reserve(Parameters["subsonic_boundary_conditions"].size());
     mSupersonicBCs.reserve(Parameters["supersonic_boundary_conditions"].size());
 
-    for(auto var_settings : Parameters["subsonic_boundary_conditions"])
+    for(const auto& var_settings : Parameters["subsonic_boundary_conditions"])
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_processes/apply_compressible_navier_stokes_boundary_conditions_process.cpp]: i am really not sure of the behaviour of Parameters if you get a reference to it. I would leave out this ones

File: applications/FluidDynamicsApplication/custom_processes/embedded_skin_visualization_process.cpp [hunk 6]
```diff
@@ -85,7 +85,7 @@ ModelPart& EmbeddedSkinVisualizationProcess::CreateAndPrepareVisualizationModelP
     // Check and add variables to visualization model part
     auto& r_origin_variables_list = r_origin_model_part.GetNodalSolutionStepVariablesList();
     auto& r_visualization_variables_list = r_visualization_model_part.GetNodalSolutionStepVariablesList();
-    for (auto var_name_param : rParameters["visualization_variables"]) {
+    for (const auto& var_name_param : rParameters["visualization_variables"]) {
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_processes/embedded_skin_visualization_process.cpp]: same, i would leave this out

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 7]
```diff
@@ -760,7 +760,7 @@ void MmgProcess<TMMGLibrary>::ApplyLocalParameters() {
     // apply (triangle or tetra), the reference of these entities and the hmin, hmax and
     // hausdorff values to apply
     for (auto parameter_settings : parameter_array) {
-        for (auto model_part_name_object : parameter_settings["model_part_name_list"])
+        for (const auto& model_part_name_object : parameter_settings["model_part_name_list"])
```
Comment (RiccardoRossi) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: same

**Review by RiccardoRossi — COMMENTED** [review]
i am not sure on the effect of taking references to paramters so i would leave that out. 

other than that, i don't see particular problems with the code we have here.
That being said, essentially all of the loops you changed are serial. in that context the pointer copy should be essentially negligible

File: applications/ContactStructuralMechanicsApplication/custom_processes/base_contact_search_process.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_processes/base_contact_search_process.cpp:747]: The reference would be always better

File: applications/FluidDynamicsApplication/custom_processes/apply_compressible_navier_stokes_boundary_conditions_process.cpp (hunk 5 above)
Comment (loumalouomega) [inline applications/FluidDynamicsApplication/custom_processes/apply_compressible_navier_stokes_boundary_conditions_process.cpp]:
```suggestion
    for(const auto var_settings : Parameters["subsonic_boundary_conditions"])
```

File: applications/FluidDynamicsApplication/custom_processes/apply_compressible_navier_stokes_boundary_conditions_process.cpp [hunk 8]
```diff
@@ -29,12 +29,12 @@ ApplyCompressibleNavierStokesBoundaryConditionsProcess::ApplyCompressibleNavierS
...
     {
         ReadBoundaryCondition(mSubsonicBCs, var_settings);
     }
 
-    for(auto var_settings : Parameters["supersonic_boundary_conditions"])
+    for(const auto& var_settings : Parameters["supersonic_boundary_conditions"])
```
Comment (loumalouomega) [inline applications/FluidDynamicsApplication/custom_processes/apply_compressible_navier_stokes_boundary_conditions_process.cpp]:
```suggestion
    for(const auto var_settings : Parameters["supersonic_boundary_conditions"])
```

File: applications/FluidDynamicsApplication/custom_processes/embedded_skin_visualization_process.cpp (hunk 6 above)
Comment (loumalouomega) [inline applications/FluidDynamicsApplication/custom_processes/embedded_skin_visualization_process.cpp]:
```suggestion
    for (const auto var_name_param : rParameters["visualization_variables"]) {
```

File: applications/FluidDynamicsApplication/custom_processes/embedded_skin_visualization_process.cpp [hunk 9]
```diff
@@ -180,7 +180,7 @@ void EmbeddedSkinVisualizationProcess::FillVariablesList(
     std::vector<const Variable<TDataType>*>& rVariablesList)
 {
     rVariablesList.clear();
-    for (auto i_var_params : rParameters) {
+    for (const auto& i_var_params : rParameters) {
```
Comment (loumalouomega) [inline applications/FluidDynamicsApplication/custom_processes/embedded_skin_visualization_process.cpp]:
```suggestion
    for (const auto i_var_params : rParameters) {
```

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp (hunk 7 above)
Comment (loumalouomega) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]:
```suggestion
        for (const auto model_part_name_object : parameter_settings["model_part_name_list"])
```

**Comment by loumalouomega** [comment]
> i am not sure on the effect of taking references to paramters so i would leave that out.
> 
> other than that, i don't see particular problems with the code we have here. That being said, essentially all of the loops you changed are serial. in that context the pointer copy should be essentially negligible

Done (reverting Parameter reference)

**Comment by RiccardoRossi** [comment]
ok then

File: applications/FluidDynamicsApplication/custom_processes/apply_compressible_navier_stokes_boundary_conditions_process.cpp [hunk 10]
```diff
@@ -29,12 +29,12 @@ ApplyCompressibleNavierStokesBoundaryConditionsProcess::ApplyCompressibleNavierS
     mSubsonicBCs.reserve(Parameters["subsonic_boundary_conditions"].size());
     mSupersonicBCs.reserve(Parameters["supersonic_boundary_conditions"].size());
 
-    for(auto var_settings : Parameters["subsonic_boundary_conditions"])
+    for(const auto var_settings : Parameters["subsonic_boundary_conditions"])
```
Comment (loumalouomega) [inline applications/FluidDynamicsApplication/custom_processes/apply_compressible_navier_stokes_boundary_conditions_process.cpp]:
```suggestion
    for(auto var_settings : Parameters["subsonic_boundary_conditions"])
```

File: applications/FluidDynamicsApplication/custom_processes/apply_compressible_navier_stokes_boundary_conditions_process.cpp [hunk 11]
```diff
@@ -29,12 +29,12 @@ ApplyCompressibleNavierStokesBoundaryConditionsProcess::ApplyCompressibleNavierS
...
     {
         ReadBoundaryCondition(mSubsonicBCs, var_settings);
     }
 
-    for(auto var_settings : Parameters["supersonic_boundary_conditions"])
+    for(const auto var_settings : Parameters["supersonic_boundary_conditions"])
```
Comment (loumalouomega) [inline applications/FluidDynamicsApplication/custom_processes/apply_compressible_navier_stokes_boundary_conditions_process.cpp]:
```suggestion
    for(auto var_settings : Parameters["supersonic_boundary_conditions"])
```

File: applications/FluidDynamicsApplication/custom_processes/embedded_skin_visualization_process.cpp [hunk 12]
```diff
@@ -85,7 +85,7 @@ ModelPart& EmbeddedSkinVisualizationProcess::CreateAndPrepareVisualizationModelP
     // Check and add variables to visualization model part
     auto& r_origin_variables_list = r_origin_model_part.GetNodalSolutionStepVariablesList();
     auto& r_visualization_variables_list = r_visualization_model_part.GetNodalSolutionStepVariablesList();
-    for (auto var_name_param : rParameters["visualization_variables"]) {
+    for (const auto var_name_param : rParameters["visualization_variables"]) {
```
Comment (loumalouomega) [inline applications/FluidDynamicsApplication/custom_processes/embedded_skin_visualization_process.cpp]:
```suggestion
    for (auto var_name_param : rParameters["visualization_variables"]) {
```

File: applications/FluidDynamicsApplication/custom_processes/embedded_skin_visualization_process.cpp [hunk 13]
```diff
@@ -180,7 +180,7 @@ void EmbeddedSkinVisualizationProcess::FillVariablesList(
     std::vector<const Variable<TDataType>*>& rVariablesList)
 {
     rVariablesList.clear();
-    for (auto i_var_params : rParameters) {
+    for (const auto i_var_params : rParameters) {
```
Comment (loumalouomega) [inline applications/FluidDynamicsApplication/custom_processes/embedded_skin_visualization_process.cpp]:
```suggestion
    for (auto i_var_params : rParameters) {
```

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 14]
```diff
@@ -760,7 +760,7 @@ void MmgProcess<TMMGLibrary>::ApplyLocalParameters() {
     // apply (triangle or tetra), the reference of these entities and the hmin, hmax and
     // hausdorff values to apply
     for (auto parameter_settings : parameter_array) {
-        for (auto model_part_name_object : parameter_settings["model_part_name_list"])
+        for (const auto model_part_name_object : parameter_settings["model_part_name_list"])
```
Comment (loumalouomega) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]:
```suggestion
        for (auto model_part_name_object : parameter_settings["model_part_name_list"])
```

**Comment by loumalouomega** [comment]
Reapproval please

**Comment by loumalouomega** [comment]
> Reapproval please

@AlejandroCornejo ?

**Review by AlejandroCornejo — APPROVED** [review]

File: applications/MeshingApplication/custom_processes/multiscale_refining_process.cpp [hunk 15]
```diff
@@ -391,7 +391,7 @@ void MultiscaleRefiningProcess::CloneNodesToRefine(IndexType& rNodeId)
 
     // Adding the nodes to the refined sub model parts
     StringVectorType sub_model_part_names = mrCoarseModelPart.GetSubModelPartNames();
-    for (auto name : sub_model_part_names)
+    for (const auto& name : sub_model_part_names)
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/multiscale_refining_process.cpp:394]: FYI C++17 added string_view, maybe useful in such cases 
Not sure as I haven't used it myself yet

**Pull request merged**